### ISSUE TITLE
Dataloader stuck on multiple GPUs 

### ISSUE THREAD

**Issue by kuri54** [huggingface/accelerate#1980]
### System Info

```Shell
- `Accelerate` version: 0.21.0 -> update 0.23.0
- Platform: Linux-6.2.0-32-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Numpy version: 1.24.4
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 251.52 GB
- GPU type: NVIDIA RTX A6000
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

```python
・
・
・

model = CLIPModel.from_pretrained('openai/clip-vit-base-patch32')
processor = AutoProcessor.from_pretrained('openai/clip-vit-base-patch32')

・
・
・
my dataset, dataloader

def train_loop(train_loader, model, optimizer, num_dataset):
    model.train()

    train_running_loss = 0.0
    all_train_running_loss = []

    for inputs in train_loader:
        optimizer.zero_grad()

        outputs = model(pixel_values=inputs['pixel_values'].squeeze(),
                                     input_ids=inputs['input_ids'].squeeze(),
                                     attention_mask=inputs['attention_mask'].squeeze(),
                                     return_loss=True
                                     )

        train_loss = outputs.loss

        accelerator.backward(train_loss)
        optimizer.step()

        train_running_loss += train_loss * inputs['pixel_values'].squeeze().size(0)

    all_train_running_loss.append(accelerator.gather(train_running_loss))

    if torch.cuda.device_count() > 1:
        all_train_running_loss = torch.cat(all_train_running_loss)[:num_dataset]

    train_loss = sum(all_train_running_loss) / num_dataset

    return train_loss

def valid_loop(valid_loader, model, num_dataset):
    model.eval()

    valid_running_loss = 0.0
    all_valid_running_loss = []

    for inputs in valid_loader:  <- From the second epoch onward, they are stacked here.
        with torch.no_grad():
            outputs = model(pixel_values=inputs['pixel_values'].squeeze(),
                            input_ids=inputs['input_ids'].squeeze(),
                            attention_mask=inputs['attention_mask'].squeeze(),
                            return_loss=True
                            )

            valid_loss = outputs.loss
            valid_running_loss += valid_loss * inputs['pixel_values'].squeeze().size(0)

    all_valid_running_loss.append(accelerator.gather(valid_running_loss))

    if torch.cuda.device_count() > 1:
        all_valid_running_loss = torch.cat(all_valid_running_loss)[:num_dataset]

    valid_loss = sum(all_valid_running_loss) / num_dataset

    return valid_loss
```

### Expected behavior

It Should end normally as if it was run on a single GPU.

I am trying to get CLIP to fine-tuning with my code.
When I tried to run it on multiple GPUs using accelerate, I encountered the following phenomenon.

- Validation dataloader stuck in 2nd epoch only on multi-GPU
Specifically, when the "for inputs in valid_loader:" process is finished, it does not proceed to the next step. train_loader process is completed. Also, both train and valid are working correctly in the first epoch.
The accelerate command at that time is as follows.
`accelerate launch --multi_gpu --num_processes=2 {script_name.py} {--arg1} {--arg2} ...`

- This will not happen when single GPU is used.
`CUDA_VISIBLE_DEVICES="0" accelerate launch {script_name.py} --arg1 --arg2 ...`

- Setting `num_workers=0` in dataloader did not change the result.

- In that case, both CPU and GPU utilization is 100%.

**Comment by muellerzr** [comment]
Are you using accelerator.prepare here at all?

**Comment by kuri54** [comment]
Used in accordance with official guidelines.
```python
model, optimizer, train_loader, valid_loader = accelerator.prepare(model,
                                                                   optimizer,
                                                                   train_loader,
                                                                   valid_loader
                                                                   )
```

**Comment by kuri54** [comment]
Addition
It is confirmed that the process inside dataloader is done.

**Comment by muellerzr** [comment]
Can you try upgrading to the latest version?

**Comment by kuri54** [comment]
Sure enough, it was not the latest version.
I updated to the latest version and tried again, but the result remained the same.

**Comment by muellerzr** [comment]
And how are you creating your dataloader and dataset? I need more with this to know just why it's getting stuck. If there's a small subsample of data you can create we can use also, this would be most helpful. Otherwise I'm essentially flying blind

**Comment by kuri54** [comment]
### model and processor
```python
model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
processor = AutoProcessor.from_pretrained("openai/clip-vit-base-patch32")
```

### dataset
```python
class CLIPDataset(Dataset):
    def __init__(self, input_dir, df, preprocess, template):
        self.dir = input_dir
        self.df = df
        self.preprocess = preprocess
        self.templates = template

    def __len__(self):
        return len(self.df.path.values)

    def __getitem__(self, idx):
        img_path = f'{self.dir}/{self.df.path.values[idx]}'

        caption = self.template.format(self.df.label.values[idx])
        image = self.preprocess(Image.open(img_path))

        return image, caption
```

A template might look something like this, for example
`tmp = 'A photo of a {}.'`

### create dataset and dataloader
```python
train_dataset = CLIPDataset(input_dir, train_df, preprocess, template)
valid_dataset = CLIPDataset(input_dir, valid_df, preprocess, template)

train_loader = torch.utils.data.DataLoader(train_dataset,
                                           batch_size=args.batch_size,
                                           shuffle=True,
                                           pin_memory=True,
                                           prefetch_factor=32,
                                           num_workers=24
                                           )

valid_loader = torch.utils.data.DataLoader(valid_dataset,
                                           batch_size=args.batch_size,
                                           pin_memory=True,
                                           prefetch_factor=32,
                                           num_workers=24
                                           )
```

**Comment by muellerzr** [comment]
@kuri54 a few things that would be helpful:

How many batches are in your dataset? What happens if you set `num_workers=0` in the `DataLoaders`? And what is your batch size? 

Also, if there is a subsample/example dataset you can provide that is similar to what you are doing, that would be most helpful

**Comment by kuri54** [comment]
@muellerzr 
I am very sorry for not informing you. Resolved.

First of all, setting num_workers=0 in DataLoaders did not change the problem, it was still stuck. I tried various other remedies that I have read about on the net, but in the end, there was no improvement.

In conclusion, the way Dataloader was made was wrong. It was so long ago that I don't remember the details, but multiple Dataloaders were created in the training loop, which caused it to get stuck.

Now it is running without any problems.
Thanks for your kind support.

This issue is closed.
